VirtoCommerce.Platform.Modules/LoadContextAssemblyResolver.cs

```diff
@@ -1,122 +1,191 @@
 using System;
 using System.Collections.Generic;
 using System.IO;
-using System.Linq;
 using System.Reflection;
 using System.Runtime.Loader;
-using Microsoft.Extensions.DependencyModel;
-using Microsoft.Extensions.DependencyModel.Resolution;
 using Microsoft.Extensions.Logging;
 using VirtoCommerce.Platform.Core.Modularity;
-using VirtoCommerce.Platform.Core.Common;
+using VirtoCommerce.Platform.Core.Modularity.Exceptions;
+using VirtoCommerce.Platform.Modules.AssemblyLoading;
+
 namespace VirtoCommerce.Platform.Modules
 {
 
     public class LoadContextAssemblyResolver : IAssemblyResolver
     {
         private readonly ILogger<LoadContextAssemblyResolver> _logger;
+        private readonly Dictionary<string, Assembly> _loadedAssemblies = new Dictionary<string, Assembly>();
+        private string _basePath;
+        private readonly HashSet<string> _additionalProbingPaths = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
         public LoadContextAssemblyResolver(ILogger<LoadContextAssemblyResolver> logger)
         {
             _logger = logger;
         }
 
         /// <summary>
-        /// Registers the specified assembly and resolves the types in it when the AppDomain requests for it.
+        /// Loads specified assembly and all its nested dependencies to default AssemblyLoadContext (ALC).
         /// </summary>
-        /// <param name="assemblyFilePath">The path to the assembly to load in the LoadFrom context.</param>    
-        public Assembly LoadAssemblyFrom(string assemblyFilePath)
+        /// <param name="assemblyPath">The path to the assembly to load.</param>
+        /// <exception cref="ModuleInitializeException">If cannot load assembly or its dependencies.</exception>
+        public Assembly LoadAssemblyFrom(string assemblyPath)
         {
-            Uri assemblyUri = GetFileUri(assemblyFilePath);
+            var assemblyUri = GetFileUri(assemblyPath);
 
             if (assemblyUri == null)
             {
-                throw new ArgumentException("The argument must be a valid absolute Uri to an assembly file.", nameof(assemblyFilePath));
+                throw new ArgumentException("The argument must be a valid absolute Uri to an assembly file.", nameof(assemblyPath));
             }
 
             if (!File.Exists(assemblyUri.LocalPath))
             {
                 throw new FileNotFoundException(assemblyUri.LocalPath);
             }
 
-            var assembly = LoadWithAllReferencedAssebliesRecursive(assemblyUri.LocalPath);
+            // Fill _basePath and _additionalProbingPaths based on assemblyPath
+            // In fact we could add some _additionalProbingPaths to support assembly storages like NuGet package cache (), if we would need
+            var assemblyDirectory = Path.GetDirectoryName(assemblyUri.LocalPath);
+            _basePath = assemblyDirectory;
```

Comment:
It is not really good to change object field inside read only functions, could we avoid this?